File: app/src/main/java/ru/job4j/todolist/store/FileStore.java
package ru.job4j.todolist.store;

import android.content.Context;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.File;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.PrintWriter;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.stream.Collectors;

import ru.job4j.todolist.Task;

public class FileStore implements IStore{
    private static FileStore INST;
    private int counter = 0;
    private Context context;
    private List<Task> tasks = new ArrayList<>();

    private FileStore(Context context) {
        this.context = context;
        File fileDir = new File(String.valueOf(context.getFilesDir()));
        File[] allFiles = fileDir.listFiles();
        for (int index = 0; index < Objects.requireNonNull(allFiles).length; index++) {
            try (BufferedReader in = new BufferedReader(new FileReader(allFiles[index]))) {
                Task task = new Task(Integer.parseInt(in.readLine()), in.readLine(), in.readLine(),in.readLine(),in.readLine());
                tasks.add(task);
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
        if (allFiles.length != 0) counter = Integer.parseInt(allFiles[allFiles.length-1].getName().split("\\.")[0]) + 1;
    }

    public static FileStore getInstance(Context context) {
        if (INST == null) {
            INST = new FileStore(context);
        }
        return INST;
    }

    @Override
    public List<Task> getTasks() {
        return tasks;
    }

    @Override
    public List<Task> getFilteredTasks(String filter) {
        return this.tasks.stream().filter(
                task -> task.getName().toLowerCase().contains(filter.toLowerCase()) ||
                        task.getDesc().toLowerCase().contains(filter.toLowerCase()) ||
                        task.getCreated().toLowerCase().contains(filter.toLowerCase()))
                .collect(Collectors.toList());
    }

    @Override
    public void addTask(String name, String description, String created) {
        File file = new File(context.getFilesDir(), (counter) + ".txt");
        try (PrintWriter out = new PrintWriter(new BufferedWriter(new FileWriter(file)))) {
            out.println(counter);
            out.println(name);
            out.println(description);
            out.println(created);
            tasks.add(new Task(counter++, name, description, created, null));
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    @Override
    public void editTask(int id, String name, String description, String closed) {
        File file = new File(context.getFilesDir(), id + ".txt");
        Task task = findTaskByID(id);
        try (PrintWriter out = new PrintWriter(new BufferedWriter(new FileWriter(file)))) {
            out.println(task.getId());
            out.println(name);
            out.println(description);
            out.println(task.getCreated());
            out.println(closed);
            task.setName(name);
            task.setDesc(description);
            task.setClosed(closed);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    @Override
    public void closeOrReopenTask(int id, String closed) {
        File file = new File(context.getFilesDir(), id + ".txt");
        Task task = findTaskByID(id);
        try (PrintWriter out = new PrintWriter(new BufferedWriter(new FileWriter(file)))) {
            out.println(task.getId());
            out.println(task.getName());
            out.println(task.getDesc());
            out.println(task.getCreated());
            out.println(closed);
            task.setClosed(closed);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    @Override
    public int deleteTask(int id) {
        File file = new File(context.getFilesDir(), id + ".txt");
        file.delete();
        Task task = findTaskByID(id);
        int result = tasks.indexOf(task);
        tasks.remove(result);
        return result;
    }

    @Override
    public void deleteAllTasks() {
        File fileDir = new File(String.valueOf(context.getFilesDir()));
        File[] allFiles = fileDir.listFiles();
        for (File file : Objects.requireNonNull(allFiles)) {
            file.delete();
        }
        tasks.clear();
    }

    @Override
    public Task findTaskByID(int id) {
        return tasks.get(getPositionOfTaskById(id));
    }

    @Override
    public int getPositionOfTaskById(int id) {
        for (int index = 0 ; index < tasks.size(); index ++) {
            if (tasks.get(index).getId() == id) {
                return index;
            }
        }
        return -1;
    }
}

File: app/src/main/java/ru/job4j/todolist/store/FileStore.kt
package ru.job4j.todolist.store

import android.content.Context
import ru.job4j.todolist.Task
import java.io.*
import java.util.*
import java.util.stream.Collectors

class FileStore private constructor(private val context: Context) : IStore {
    private var counter = 0
    private val tasks: MutableList<Task> = ArrayList()

    private fun updateStore() {
        val allFiles = File(context.filesDir.toString()).listFiles()
        allFiles?.forEach {
            BufferedReader(FileReader(it)).use { `in` ->
                val task = Task(`in`.readLine().toInt(), `in`.readLine(),
                        `in`.readLine(), `in`.readLine(), `in`.readLine(), `in`.readLine())
                if (task.closed == "") task.closed = null
                tasks.add(task)
            }
        }
        if (allFiles != null && allFiles.isNotEmpty()) {
            counter = allFiles[allFiles.size - 1].name.split("\\.".toRegex()).toTypedArray()[0].toInt() + 1
        }
    }

    override fun getTasks(): List<Task> = tasks

    override fun getFilteredTasks(filter: String): List<Task> =
            tasks.stream().filter { (_, name, desc, created) ->
                name.toLowerCase().contains(filter.toLowerCase()) ||
                        desc.toLowerCase().contains(filter.toLowerCase()) ||
                        created.toLowerCase().contains(filter.toLowerCase())
            }.collect(Collectors.toList())

    override fun addTask(name: String, description: String, created: String, photo: String?) {
        val file = File(context.filesDir, "$counter.txt")
        PrintWriter(BufferedWriter(FileWriter(file))).use { out ->
            out.println(counter)
            out.println(name)
            out.println(description)
            out.println(created)
            out.println("")
            out.println(photo)
            tasks.add(Task(counter++, name, description, created, null, photo))
        }
    }

    override fun editTask(id: Int, name: String, description: String, closed: String, photo: String) {
        val file = File(context.filesDir, "$id.txt")
        findTaskByID(id)?.apply {
            PrintWriter(BufferedWriter(FileWriter(file))).use { out ->
                out.println(this.id)
                out.println(name)
                out.println(description)
                out.println(created)
                out.println(closed)
                out.println(photo)
                this.name = name
                this.desc = description
                this.closed = closed
                this.photo = photo
            }
        }

    }

    override fun closeOrReopenTask(id: Int, closed: String?) {
        val file = File(context.filesDir, "$id.txt")
        findTaskByID(id)?.apply {
            PrintWriter(BufferedWriter(FileWriter(file))).use { out ->
                out.println(this.id)
                out.println(name)
                out.println(desc)
                out.println(created)
                out.println(closed)
                out.println(photo)
                this.closed = closed
            }
        }
    }

    override fun deleteTask(id: Int): Int =
            tasks.indexOf(findTaskByID(id)).apply {
                File(context.filesDir, "$id.txt").delete()
                tasks.removeAt(this)
            }


    override fun deleteAllTasks(): Unit =
            File(context.filesDir.toString()).listFiles()?.let {
                it.forEach { file -> file.delete() }
                tasks.clear()
            }!!

    override fun findTaskByID(id: Int): Task? = tasks.firstOrNull { task -> task.id == id }

    companion object {
        private lateinit var INST: FileStore
        fun getInstance(context: Context): FileStore {
            if (!this::INST.isInitialized) {
                INST = FileStore(context)
                INST.updateStore()
            }
            return INST
        }
    }
}